File: app/schemas/task.py
from pydantic import BaseModel, Field
from datetime import datetime
from typing import Optional

class TaskBase(BaseModel):
    title: str = Field(..., min_length=1, max_length=200)
    description: Optional[str] = None
    status: str = Field(default="new", pattern="^(new|in_progress|done|cancelled)$")
    priority: str = Field(default="medium", pattern="^(low|medium|high)$")
    assignee: Optional[str] = None

class TaskCreate(TaskBase):
    pass

class TaskUpdate(BaseModel):
    title: Optional[str] = Field(None, min_length=1, max_length=200)
    description: Optional[str] = None
    status: Optional[str] = Field(None, pattern="^(new|in_progress|done|cancelled)$")
    priority: Optional[str] = Field(None, pattern="^(low|medium|high)$")
    assignee: Optional[str] = None

class TaskResponse(TaskBase):
    id: int
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True

File: client/src/types/task.ts
export const TaskStatus = {
  NEW: 'new',
  IN_PROGRESS: 'in_progress',
  DONE: 'done',
  CANCELLED: 'cancelled'
} as const

export type TaskStatus = typeof TaskStatus[keyof typeof TaskStatus]

export const TaskPriority = {
  LOW: 'low',
  MEDIUM: 'medium',
  HIGH: 'high'
} as const

export type TaskPriority = typeof TaskPriority[keyof typeof TaskPriority]

export interface TaskBase {
  title: string
  description?: string | null
  status: TaskStatus
  priority: TaskPriority
  assignee?: string | null
}

export interface TaskCreate extends TaskBase {}

export interface TaskUpdate {
  title?: string
  description?: string | null
  status?: TaskStatus
  priority?: TaskPriority
  assignee?: string | null
}

export interface TaskStatusUpdate {
  status: TaskStatus
}

export interface TaskResponse extends TaskBase {
  id: number
  created_at: string
  updated_at: string
}

export const SortBy = {
  ID: 'id',
  TITLE: 'title',
  CREATED_AT: 'created_at',
  UPDATED_AT: 'updated_at',
  PRIORITY: 'priority',
  STATUS: 'status',
  ASSIGNEE: 'assignee'
} as const

export type SortBy = typeof SortBy[keyof typeof SortBy]

export const SortOrder = {
  ASC: 'asc',
  DESC: 'desc'
} as const

export type SortOrder = typeof SortOrder[keyof typeof SortOrder]

export interface TaskFilters {
  status?: TaskStatus
  assignee?: string
  priority?: TaskPriority
  sort_by?: SortBy
  sort_order?: SortOrder
  skip?: number
  limit?: number
}
